source/material_model/rheology/diffusion_dislocation.cc

```diff
@@ -88,99 +89,90 @@ namespace aspect
-                                               "Residual is " + Utilities::to_string(log_strain_rate_residual) +
-                                               " after " + Utilities::to_string(stress_iteration) + " iterations. "
-                                               "You can increase the number of iterations by adapting the "
-                                               "parameter 'Maximum strain rate ratio iterations'."));
-
-                        log_strain_rate_residual = std::abs(std::log(diffusion_strain_rate/old_diffusion_strain_rate));
-                        stress_ii = 2.0 * edot_ii * 1./(1./diffusion_viscosity + 1./dislocation_viscosity);
-                      }
-                    while (log_strain_rate_residual > log_strain_rate_residual_threshold);
-
-                    break;
-                  }
-              }
-
-            // The effective viscosity, with minimum and maximum bounds
+            Vector<double> log_stress_ii(1);
+            log_stress_ii[0] = std::log(stress_ii);
+
+            typename SUNDIALS::KINSOL<Vector<double>>::AdditionalData additional_data;
+            additional_data.function_tolerance = log_strain_rate_residual_threshold;
+
+            SUNDIALS::KINSOL<Vector<double>> nonlinear_solver(additional_data);
+
+            nonlinear_solver.reinit_vector = [&](Vector<double> &x)
+            {
+              x.reinit(1);
+            };
+
+            nonlinear_solver.residual =
+              [&](const Vector<double> &evaluation_point,
+                  Vector<double>       &residual)
+            {
+              compute_log_strain_rate_residual(evaluation_point, residual, pressure, temperature, diffusion_creep_parameters, dislocation_creep_parameters, log_edot_ii);
+            };
+
+            nonlinear_solver.setup_jacobian =
+              [&](const Vector<double> &current_u,
+                  const Vector<double> & /*current_f*/)
+            {
+              compute_log_strain_rate_deriv(current_u, pressure, temperature, diffusion_creep_parameters, dislocation_creep_parameters, log_strain_rate_deriv);
+            };
+
+            nonlinear_solver.solve_with_jacobian = [&](const Vector<double> &rhs,
+                                                       Vector<double> &solution,
+                                                       const double /*tolerance*/)
+            {
+              solution(0) -= rhs(0)/log_strain_rate_deriv;
+            };
+
+            nonlinear_solver.solve(log_stress_ii);
+
+            stress_ii = std::exp(log_stress_ii[0]);
             composition_viscosities[j] = std::min(std::max(stress_ii/edot_ii/2, minimum_viscosity), maximum_viscosity);
           }
         return composition_viscosities;
       }
 
+
+      template <int dim>
+      void
+      DiffusionDislocation<dim>::compute_log_strain_rate_deriv(
+        const Vector<double> &evaluation_point,
```

Comment:
This will become a scalar with my suggestion above, rather than a length-1 vector. I think it would be useful to give it a more descriptive name than `evaluation_point` -- perhaps `current_log_strain_rate` or so?